### PR TITLE
MMYOLO TensorRT/ONNXRUNTIME deploy in MMDEPLOY

### PR DESCRIPTION
**Pull request by triple-Mu**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily receiving feedbacks. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.
MMYOLO TensorRT/ONNXRUNTIME deploy in MMDEPLOY

## Modification

Please briefly describe what modification is made in this PR.
Such files changes

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdeploy/codebase/__init__.py [hunk 1]
```diff
@@ -6,7 +6,8 @@
 
 extra_dependent_library = {
     Codebase.MMOCR: ['mmdet'],
-    Codebase.MMROTATE: ['mmdet']
+    Codebase.MMROTATE: ['mmdet'],
+    Codebase.MMYOLO: ['mmyolo']
```
Comment (RunningLeon) [inline mmdeploy/codebase/__init__.py:10]: mmyolo has dependency of mmdet, not mmyolo itself.

File: mmdeploy/codebase/base/task.py [hunk 2]
```diff
@@ -176,10 +176,10 @@ def _merge_cfg(cfg):
                     ' field of config. Please set '\
                     '`visualization=dict(type="VisualizationHook")`'
 
-                cfg.default_hooks.visualization.enable = True
+                cfg.default_hooks.visualization.draw = True
```
Comment (RunningLeon) [inline mmdeploy/codebase/base/task.py:179]: we better keep mmyolo same arguments as mmdet in here @hhaAndroid

File: mmdeploy/codebase/base/task.py [hunk 3]
```diff
@@ -176,10 +176,10 @@ def _merge_cfg(cfg):
...
                 cfg.default_hooks.visualization.show = show
                 cfg.default_hooks.visualization.wait_time = wait_time
-                cfg.default_hooks.visualization.out_dir = show_dir
+                cfg.default_hooks.visualization.test_out_dir = show_dir
```
Comment (RunningLeon) [inline mmdeploy/codebase/base/task.py:182]: is it really necessary to change here?

File: mmdeploy/codebase/mmyolo/deploy/model_partition_cfg.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
+# Copyright (c) OpenMMLab. All rights reserved.
```
Comment (RunningLeon) [inline mmdeploy/codebase/mmyolo/deploy/model_partition_cfg.py:1]: if mmyolo does not support onnx model partition, we can remove this config

File: mmdeploy/codebase/mmyolo/deploy/object_detection.py [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+    module_dict = DATASETS.module_dict
+
+    for dataloader_name in [
+            'test_dataloader', 'val_dataloader', 'train_dataloader'
+    ]:
+        if dataloader_name not in model_cfg:
+            continue
+        dataloader_cfg = model_cfg[dataloader_name]
+        dataset_cfg = dataloader_cfg.dataset
+        dataset_cls = module_dict.get(dataset_cfg.type, None)
```
Comment (RunningLeon) [inline mmdeploy/codebase/mmyolo/deploy/object_detection.py:86]: should give a better name to variable `dataset_cls` since it has nothing to do with `cls`

File: mmdeploy/codebase/mmyolo/deploy/object_detection.py [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+        return preprocess
+
+    def get_postprocess(self) -> Dict:
+        """Get the postprocess information for SDK.
+
+        Return:
+            dict: Composed of the postprocess information.
+        """
+        postprocess = self.model_cfg.model.test_cfg
+        if 'rpn' in postprocess:
```
Comment (RunningLeon) [inline mmdeploy/codebase/mmyolo/deploy/object_detection.py:227]: have we support sdk for mmyolo?

File: mmdeploy/codebase/mmyolo/deploy/object_detection.py [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+            postprocess['min_bbox_size'] = postprocess['rpn']['min_bbox_size']
+        if 'rcnn' in postprocess:
```
Comment (RunningLeon) [inline mmdeploy/codebase/mmyolo/deploy/object_detection.py:229]: is there any instance segmentation model in mmyolo? if not, could remove mask related configs

File: mmdeploy/codebase/base/task.py (hunk 3 above)
Comment (triple-Mu) [reply mmdeploy/codebase/base/task.py:182]: Yes. If not I can not pass mmyolo model visualization

Comment (RunningLeon) [reply mmdeploy/codebase/base/task.py:182]: then  mmyolo should change to align with most codebases @hhaAndroid

File: mmdeploy/codebase/__init__.py (hunk 1 above)
Comment (RunningLeon) [inline mmdeploy/codebase/__init__.py:10]: here mmdet

File: mmdeploy/codebase/mmyolo/__init__.py [hunk 8]
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from .deploy import (ObjectDetection, clip_bboxes, get_post_processing_params,
+                     pad_with_value, pad_with_value_if_necessary)
```
Comment (hanrui1sensetime) [inline mmdeploy/codebase/mmyolo/__init__.py:3]: Please remove `clip_bboxes`, `get_post_processing_params`, `pad_with_value`, `pad_with_value_if_necessary`, or will failed to convert model.

File: mmdeploy/codebase/mmyolo/deploy/__init__.py [hunk 9]
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from .object_detection import MMYolo, ObjectDetection
+from .utils import (clip_bboxes, get_post_processing_params, pad_with_value,
+                    pad_with_value_if_necessary)
```
Comment (hanrui1sensetime) [inline mmdeploy/codebase/mmyolo/deploy/__init__.py:4]: Please remove clip_bboxes, get_post_processing_params, pad_with_value, pad_with_value_if_necessary, or will failed to convert model.

File: mmdeploy/codebase/mmyolo/models/dense_heads/yolov5_head.py [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+def yolov5_head__predict_by_feat(ctx,
+                                 self,
+                                 cls_scores: List[Tensor],
+                                 bbox_preds: List[Tensor],
+                                 objectnesses: Optional[List[Tensor]],
+                                 batch_img_metas: Optional[List[dict]] = None,
+                                 cfg: Optional[ConfigDict] = None,
+                                 rescale: bool = False,
+                                 with_nms: bool = True) -> List[InstanceData]:
+    assert len(cls_scores) == len(bbox_preds)
```
Comment (RunningLeon) [inline mmdeploy/codebase/mmyolo/models/dense_heads/yolov5_head.py:27]: add docstring here and describe what changes

**Comment by RunningLeon** [comment]
@triple-Mu Hi, could you add docstring and unit test for this PR? Besides, docs should be updated as well.

File: mmdeploy/codebase/mmyolo/deploy/__init__.py [hunk 11]
```diff
@@ -0,0 +1,4 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from .object_detection import MMYolo, ObjectDetection
```
Comment (grimoire) [inline mmdeploy/codebase/mmyolo/deploy/__init__.py:2]: Why don't we import YoloObjectDetection instead of ObjectDetection?

File: mmdeploy/codebase/mmyolo/models/__init__.py [hunk 12]
```diff
@@ -0,0 +1,2 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from .dense_heads import *  # noqa: F401,F403
```
Comment (grimoire) [inline mmdeploy/codebase/mmyolo/models/__init__.py:2]: use `from . import dense_heads`

File: mmdeploy/codebase/mmyolo/models/dense_heads/__init__.py [hunk 13]
```diff
@@ -0,0 +1,4 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from .yolov5_head import yolov5_head__predict_by_feat
```
Comment (grimoire) [inline mmdeploy/codebase/mmyolo/models/dense_heads/__init__.py:2]: use `from . import yolov5_head`. We do not want to expose the rewriter to the module outside.

File: mmdeploy/codebase/mmyolo/models/dense_heads/yolov5_head.py [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+    flatten_cls_scores = [
+        cls_score.permute(0, 2, 3, 1).reshape(num_imgs, -1, self.num_classes)
+        for cls_score in cls_scores
+    ]
+    flatten_bbox_preds = [
+        bbox_pred.permute(0, 2, 3, 1).reshape(num_imgs, -1, 4)
+        for bbox_pred in bbox_preds
+    ]
+
+    if with_objectnesses:
```
Comment (grimoire) [inline mmdeploy/codebase/mmyolo/models/dense_heads/yolov5_head.py:61]: We can remove this flag(line 28~32), just check if objectnesses is not None here. The assert can also be placed here.

File: mmdeploy/codebase/mmyolo/models/dense_heads/yolov5_head.py [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+    'YOLOv5Head.predict_by_feat')
+def yolov5_head__predict_by_feat(ctx,
+                                 self,
+                                 cls_scores: List[Tensor],
+                                 bbox_preds: List[Tensor],
+                                 objectnesses: Optional[List[Tensor]],
+                                 batch_img_metas: Optional[List[dict]] = None,
+                                 cfg: Optional[ConfigDict] = None,
+                                 rescale: bool = False,
+                                 with_nms: bool = True) -> List[InstanceData]:
```
Comment (grimoire) [inline mmdeploy/codebase/mmyolo/models/dense_heads/yolov5_head.py:26]: The output of the rewriter is not `List[InstanceData]`.

File: mmdeploy/codebase/mmyolo/models/dense_heads/yolov5_head.py [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+        flatten_objectness = [
+            objectness.permute(0, 2, 3, 1).reshape(num_imgs, -1)
+            for objectness in objectnesses
+        ]
+    else:
+        flatten_objectness = [None for _ in range(len(featmap_sizes))]
+
+    cls_scores = torch.cat(flatten_cls_scores, dim=1).sigmoid()
+    flatten_bbox_preds = torch.cat(flatten_bbox_preds, dim=1)
+    flatten_objectness = torch.cat(flatten_objectness, dim=1).sigmoid()
```
Comment (grimoire) [inline mmdeploy/codebase/mmyolo/models/dense_heads/yolov5_head.py:71]: flatten_objectnessc could be none?

File: mmdeploy/codebase/mmyolo/models/dense_heads/yolov5_head.py [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+    bboxes = self.bbox_coder.decode(flatten_priors[None], flatten_bbox_preds,
+                                    flatten_stride)
+
+    # directly multiply score factor and feed to nms
+    scores = cls_scores * (flatten_objectness.unsqueeze(-1))
+    max_scores, _ = torch.max(scores, 1)
```
Comment (grimoire) [inline mmdeploy/codebase/mmyolo/models/dense_heads/yolov5_head.py:77]: We do not need to do the threshold here. There is a score threshold inside NMS.

File: mmdeploy/codebase/mmyolo/models/__init__.py (hunk 12 above)
Comment (triple-Mu) [reply mmdeploy/codebase/mmyolo/models/__init__.py]: Got it

File: mmdeploy/codebase/mmyolo/models/dense_heads/__init__.py (hunk 13 above)
Comment (triple-Mu) [reply mmdeploy/codebase/mmyolo/models/dense_heads/__init__.py]: Got it

File: mmdeploy/codebase/mmyolo/models/dense_heads/yolov5_head.py (hunk 15 above)
Comment (triple-Mu) [reply mmdeploy/codebase/mmyolo/models/dense_heads/yolov5_head.py]: Is it Tuple[InstanceData]?

File: mmdeploy/codebase/mmyolo/models/dense_heads/yolov5_head.py (hunk 16 above)
Comment (triple-Mu) [reply mmdeploy/codebase/mmyolo/models/dense_heads/yolov5_head.py:62]: Got it. I will delete it

File: mmdeploy/codebase/mmyolo/models/dense_heads/yolov5_head.py (hunk 17 above)
Comment (triple-Mu) [reply mmdeploy/codebase/mmyolo/models/dense_heads/yolov5_head.py]: Got it. I will delete it

File: mmdeploy/codebase/mmyolo/models/dense_heads/yolov5_head.py (hunk 10 above)
Comment (triple-Mu) [reply mmdeploy/codebase/mmyolo/models/dense_heads/yolov5_head.py:55]: DONE

File: mmdeploy/codebase/mmyolo/models/dense_heads/yolov5_head.py (hunk 15 above)
Comment (grimoire) [reply mmdeploy/codebase/mmyolo/models/dense_heads/yolov5_head.py]: Tuple[Tensor]
